### ISSUE TITLE
Encrypted metrics

### ISSUE THREAD

**Issue by kyessenov** [istio/istio#50114]
Plaintext Prometheus metrics violate the compliance requirements for the regulated users. Istio should provide a mode where 1) metrics are encrypted; 2) encryption is FIPS compliant.

**Comment by hzxuzhonghu** [comment]
hmm, does kubernetes support encrypted metrics, i donot think so

**Comment by kyessenov** [comment]
Good question @hzxuzhonghu . Istio metrics contain identifiable information for the mesh workloads, e.g. you can infer which workloads are in the cluster, what labels they have, and how much load they take. This is not possible on k8s without RBAC permissions. Most of the traditional metrics are operational and low-information, so they do not pose information leakage concerns, similar to how health probes are not sensitive.

**Comment by zirain** [comment]
cc @leosarra

**Comment by leosarra** [comment]
Hello @kyessenov
I agree that it would be great to have metrics exposed via a secured server.
Ideally I was thinking of allowing users to enable a secured server and provide their own certs via a secret, but I'm afraid that it would create some pain points regarding UX.
For example, do we think it is reasonable to expect users to create the secret holding the certs in every namespace where a sidecar/controller could be deployed?

Anyway, I was planning to make an RFC document after Easter so that we can discuss the approach and determine whether or not to move forward with it.

**Comment by hzxuzhonghu** [comment]
Or instead, istio can export/push the metrics.  It should know where the secure target is, whether via HTTP or https

**Comment by ksubrmnn** [comment]
Hi folks,

I've been looking into this myself with Prometheus + Istio following [these docs](https://istio.io/latest/docs/ops/integrations/prometheus/#tls-settings). The user experience is pretty painful as it is and the documentation can be a bit hard to follow for someone new.

I'd love to help out on this work item

**Comment by kyessenov** [comment]
@ksubrmnn Right, Istio mTLS is not a good solution with Prometheus because of two issues:
1) Permissive mode requires Prometheus to speak Istio mTLS which is not standard mTLS. 
2) Applying mTLS requires "service addressing" from Prometheus to every workload in "strict mode". Normally, pod-to-pod communication is plaintext in Istio and you have to use "service" to apply mTLS, but that's not how Prometheus works. 

One potential option is to use a collector sidecar to convert to push-based collection, which doesn't suffer from 2).

**Comment by hemendrateli** [comment]
Hi All,
I work with @kyessenov and working on implementing this.

There are total 3 kinds of metrics :- application metrics, sidecar metrics and control plane metrics.

When we use a collector sidecar then a collector sidecar can listen to localhost and encrypt the metrics before sending it to Prometheus. In this case we need to block plain text access to metrics endpoints.

For application metrics we can have one PeerAuthentication policy to force all traffic to use strict mTLS, but also excluding the encrypted metrics port of the collector sidecar.

For proxy and control plane metrics we need to block the access from outside of the pod so that no one can access from outside. 
We are proposing to have an environment variable for Pilot and Proxy, there would be http middleware handler for metrics handlers which would check if env variable is set then accept request from localhost only. We can't restrict host to localhost for endpoint for http server because same endpoint is used for debug and health also.

**Comment by hemendrateli** [comment]
Created https://github.com/istio/istio/pull/50728 for to add support for disabling metrics endpoint from outside of pod.

**Comment by howardjohn** [comment]
Closing since https://github.com/istio/istio/pull/50728 merged

**Comment by leosarra** [comment]
@hemendrateli If I understand correctly, we currently have a method to not expose the metric endpoint outside (#50728) but still no support for a collector sidecar. Should we reopen this pull request?

**Comment by hemendrateli** [comment]
Yes, still no support for a collector sidecar.

**Comment by kfaseela** [comment]
> Hello [@kyessenov](https://github.com/kyessenov) I agree that it would be great to have metrics exposed via a secured server. Ideally I was thinking of allowing users to enable a secured server and provide their own certs via a secret, but I'm afraid that it would create some pain points regarding UX. For example, do we think it is reasonable to expect users to create the secret holding the certs in every namespace where a sidecar/controller could be deployed?
> 
> Anyway, I was planning to make an RFC document after Easter so that we can discuss the approach and determine whether or not to move forward with it.

Should we reopen this issue?

**Comment by dkrieger** [comment]
please reopen this issue

I'll also note that it makes a lot more sense for istio to support in-mesh metric proxies that communicate w/ istiod and envoy proxies via TLS (ideally mtls but involving cert manager et al is better than nothing) and let prometheus scraping occur entirely in-mesh. it seems with ambient/ztunnel this should be doable

observability and security shouldn't be at odds like this in a service mesh that is meant to provide both

**Comment by frittentheke** [comment]
Related to this issue of providing metrics via an encrypted (and authenticated?) endpoint is the discussion around compression and a change in how metrics are exposed: https://github.com/istio/istio/issues/58697#issuecomment-3812203409

If Envoy was used to expose all of the metrics (including agent metrics), there likely is a good chance TLS and authentication could be added quite easily?

**Comment by kfaseela** [comment]
A solution was implemented here: https://github.com/istio/istio/pull/60391
